### ISSUE TITLE
MaxPool2D Returns Wrong Shape With Ceil_Mode

### ISSUE THREAD

**Issue by AndrewZhaoLuo** [pytorch/pytorch#57178]
## 🐛 Bug

The documentation is clear on the output shape given parameters to the function:

https://pytorch.org/docs/stable/generated/torch.nn.MaxPool2d.html

However the behavior does not match the documentation.

## To Reproduce

Run
```
import torch

out = torch.nn.MaxPool2d(kernel_size=[1, 1], dilation=[1, 1], stride=[2, 2], ceil_mode=True)(
            torch.rand([1, 3, 32, 32]).float()
        )

print(out.shape)
```

>>  torch.Size([1, 3, 16, 16])

## Expected behavior

By my calculation the result should be torch.Size([1, 3, 17, 17]) based on the documentation. 

For example for H_out, plug in padding[0] = 0, dilation=1, kernel_size[0] = 1, stride[0] = 2.

## Environment

Collecting environment information...
PyTorch version: 1.8.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 11.2.3 (arm64)
GCC version: Could not collect
Clang version: 11.1.0
CMake version: version 3.19.7

Python version: 3.8 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.20.1
[pip3] torch==1.8.0
[pip3] torchvision==0.2.2
[conda] numpy                     1.20.1           py38h9e6c65a_0    conda-forge
[conda] pytorch                   1.8.0           cpu_py38hc43b888_0    conda-forge
[conda] torchvision               0.2.2                      py_3    pytorch

**Comment by eqy** [comment]
Can you clarify how you get 17? With a kernel size of [1,1] and stride of [2,2] the first spatial output [0,0] comes from input [0,0] (the second output [0, 1] from input [0, 2]) increasing until we get to output [0, 15] and input [0, 30]. Note that [0,32] is outside of the bounds of the input and since it is the _start_ of the next location, `ceil_mode` does not affect the computation here. The same applies for the vertical spatial dimension for a spatial output size of [16, 16].

**Comment by AndrewZhaoLuo** [comment]
Yeah it makes sense why it would be [16, 16], there probably needs to be some clarification on the behavior. Also, probably good to check onnxruntime for this edge case too. It has similar claimed behavior: https://github.com/onnx/onnx/blob/master/docs/Operators.md#MaxPool. 

I personally can't try onnx since I only have access to an m1 mac and onnxruntime doesn't support m1 macs right now.

**Comment by xwang233** [comment]
Likely a known issue. See https://github.com/pytorch/pytorch/issues/45357 https://github.com/pytorch/pytorch/pull/46558

cc @heitorschueroff

**Comment by heitorschueroff** [comment]
Elevating this to high priority due to the number of reported issues.

**Comment by eqy** [comment]
Do we know if the correct behavior is defined by the current implementation or the equation shape equation in the docs?

**Comment by heitorschueroff** [comment]
@eqy As you correctly noted in your comment above, the equation in the documentation has one corner case described in the following note from the doc

```
When ceil_mode=True, sliding windows are allowed to go off-bounds if they start within the left padding or the input. Sliding windows that would start in the right padded region are ignored.
```

Which essentially says that we ignore sliding windows that do not start within the left padding or the input. The documentation in this case is correct albeit confusing. The implementation is also not thoroughly tested so I think it's worth revising it.

**Comment by eqy** [comment]
A slight clarification I think it could be worth making is that "sliding windows are allowed to go off-bounds if they start within the left padding or the input _and_ _start_ _from_ _a_ _previously_ _unused_ _input_ _location_." Otherwise, the language is ambiguous as to whether e.g., kernel_size 2 and stride 1 yields a 1x1 or 2x2 output when given a 2x2 input, as starting a window with top left corner at [1,1] is valid according to the current phrasing.
